### PR TITLE
Citus standby replication

### PR DESCRIPTION
**Pull request by gpacix**

### CODE REVIEW

File: deploy/roles/odbc/templates/odbcinst.ini [hunk 1]
```diff
@@ -30,7 +30,7 @@ useNTLMv2               = True
 
 {% for driver in odbcinst_entries %}
 [{{ driver.drivername }}]
-{% for key, value in driver.attributes.iteritems() %}
+{% for key, value in driver.attributes.items() %}
```
Comment (dvbalen) [inline deploy/roles/odbc/templates/odbcinst.ini:33]: {% for key, value in list( driver.attributes.items() ) %}

if you want it to work in both 2 and 3

File: deploy/roles/psql-authnz/tasks/main.yml [hunk 2]
```diff
@@ -5,18 +5,7 @@
 
 - name: Install dependencies
   yum:
-    name: "{{item}}"
+    name: ['python-devel', 'libevent-devel', 'gcc', 'gcc-c++', 'kernel-devel', 'libxslt-devel', 'libffi-devel', 'openssl-devel', 'openldap-devel']
```
Comment (dvbalen) [inline deploy/roles/psql-authnz/tasks/main.yml]: For these long lists an item per line would be more readable

File: .devcontainer/Makefile [hunk 3]
```diff
@@ -0,0 +1,15 @@
+this_container=$(shell docker ps --filter "name=${CONTAINERNAME}" -aq)
+devcontainers=$(shell docker ps --filter "name=devcontainer" -aq | grep -v ${this_container})
+
+buildbase:
+	cd ${PROJECT_ROOT}/.devcontainer && docker build -f Dockerfile-base --rm --no-cache -t centos7-sshd .
+deploy:
+	ansible-playbook ${PROJECT_ROOT}/deploy/deploy_citus_server.yml -i ${PROJECT_ROOT}/deploy/vagrant_hosts \
+	--ask-pass --user=super --ask-sudo-pass
```
Comment (tanderegg) [inline .devcontainer/Makefile:8]: This is more of a big picture comment, and doesn't need to be addressed in this PR, but should we look into using the Docker provider for Vagrant rather than a Makefile?  Otherwise we will basically need to recreate the Vagrantfile as a Makefile, with build steps for each playbook.

File: .devcontainer/bashrc [hunk 4]
```diff
@@ -0,0 +1,20 @@
+export WORKSPACE=${WORKSPACE:-/workspace}
+cat $WORKSPACE/.devcontainer/welcome.txt
+
+alias deploy='ansible-playbook $WORKSPACE/deploy/deploy_nginx_proxy_server.yml -i $WORKSPACE/deploy/vagrant_hosts --ask-pass --user=super --ask-sudo-pass'
```
Comment (tanderegg) [inline .devcontainer/bashrc:4]: Might help to document what these commands are for, for example why is the Nginx proxy server called out explicitly here?

File: .devcontainer/docker-compose.yml [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+      # security_opt:
+      #   - seccomp:unconfined
+  
+      # Overrides default command so things don't shut down after the process ends.
+      command: sleep infinity 
+      networks:
+        ansible_nw:
+          ipv4_address: 10.0.1.253
+  #please note: the spacing needs to line up w/ the other yaml templates
+    citus_coordinator:
```
Comment (tanderegg) [inline .devcontainer/docker-compose.yml:43]: This file also seems like another approach to doing what Vagrant does.  It is it easy with docker-compose to only spin up a few of the listed containers, and restart them individually?  Again, it might be easier to rely on Vagrant/Docker rather than create a new workflow.

File: .travis.yml [hunk 6]
```diff
@@ -7,167 +7,189 @@ env:
   - container_id=$(mktemp)
 
  matrix:
-#  proxy server
+# 1  proxy server
   - >
+    JOB_NAME="Proxy Server"
```
Comment (tanderegg) [inline .travis.yml:12]: :+1: More clarity is always helpful!

File: archive-wal-file [hunk 7]
```diff
@@ -0,0 +1,32 @@
+#!/usr/bin/env bash
```
Comment (tanderegg) [inline archive-wal-file]: I think this might have been included accidentally, only the template in the `citus_coordinator` role is needed, right?

File: deploy/deploy_admin_terminal.yml [hunk 8]
```diff
@@ -28,5 +28,5 @@
         name: ansible
         state: present
         umask: "0022"
-      sudo: no
+      #sudo: no
```
Comment (tanderegg) [inline deploy/deploy_admin_terminal.yml]: Did the `sudo: no` cause problems?  In generally the goal was just to ensure that ansible installed correctly, and allow users to deploy from the Admin Terminal manually  if necessary (we rarely if ever do, however).

File: deploy/deploy_citus_server.yml [hunk 9]
```diff
@@ -1,6 +1,13 @@
 # deploy_citus_server.yml
 ---
-- hosts: citus
+- name: Install common and optional packages, JDK, Postgresql, and Citus
```
Comment (tanderegg) [inline deploy/deploy_citus_server.yml:3]: :+1:

File: deploy/deploy_citus_server.yml [hunk 10]
```diff
@@ -1,6 +1,13 @@
...
+  hosts: citus
+  # Avoid nasty surprises when someone does "--extra-vars pp_install=False", etc.:
+  pre_tasks:
+  - set_fact:
+      install_pp: "{{ pp_install | default(False) | bool }}"
```
Comment (tanderegg) [inline deploy/deploy_citus_server.yml]: Are these essentially to sanitize the value?  Is it possible to use `pp_install:` at the beginning rather than `install_pp`, or does that not work?  I'm asking only because having two different named variables that are meant to do the same thing could be confusing.

File: deploy/deploy_citus_server.yml [hunk 11]
```diff
@@ -103,15 +119,104 @@
...
-- hosts: citus_coordinator_standby
+- name: Set up Citus Coordinator Standby
+  hosts: citus_coordinator_standby
+  pre_tasks:
+  - set_fact:
+      install_pp: "{{ pp_install | default(False) | bool }}"
   serial: 1
   roles:
-    - {role: citus_coordinator_standby, when: "{{ pp_install }}"}
+    - {role: citus_coordinator_standby, when: "pp_install | default(False) | bool"}
```
Comment (tanderegg) [inline deploy/deploy_citus_server.yml:156]: Should we use `install_pp` here rather than re-sanitizing?

File: deploy/deploy_citus_server.yml [hunk 12]
```diff
@@ -103,15 +119,104 @@
...
+        # TODO: make this a variable
+        path: /var/lib/pgsql/.ssh/id_rsa
+        comment: "key for postgres@{{ ansible_hostname }}"
+        owner: postgres
+        group: postgres
+        mode: "u=rw,go="
+        state: present # this is the default anyway
+      register: coord_key
+  
+    - name: Fix owner of the generated pub key
```
Comment (tanderegg) [inline deploy/deploy_citus_server.yml:190]: Annoying that it doesn't use the right values for both files!

File: deploy/deploy_citus_server.yml [hunk 13]
```diff
@@ -103,15 +119,104 @@
...
+      file:
+        path: /var/lib/pgsql/.ssh/id_rsa.pub
+        owner: postgres
+        group: postgres
+        mode: "u=rw,go=r"
+
+    - name: Add standby's hostkey to postgres@coordinator
+      lineinfile:
+        path: "~postgres/.ssh/known_hosts"
+        insertafter: "{{hostvars[groups['citus_coordinator_standby'][0]].ansible_ssh_host_key_ecdsa_public}}"
```
Comment (tanderegg) [inline deploy/deploy_citus_server.yml]: Minor nitpick, but generally I try to leave spaces between the curly braces and their content for readability.

File: deploy/deploy_citus_server.yml [hunk 14]
```diff
@@ -103,15 +119,104 @@
...
+    - name: Create .ssh directory for user postgres
+      file:
+        state: directory
+        path: "/var/lib/pgsql/.ssh"
+        owner: postgres
+        group: postgres
+        mode: "u=rwx,go="
+
+    - name: Make postgres@standby trust public key of postgres@coordinator
+      ## delegate_to: central_server_name ## what's this?
```
Comment (tanderegg) [inline deploy/deploy_citus_server.yml]: Not sure what the question is here, `delegate_to` allows the Ansible command to run on a different server with reference to the host in question.

File: deploy/deploy_citus_server.yml [hunk 15]
```diff
@@ -103,15 +119,104 @@
...
+      # TODO: add a "when: coord_key.changed
+      authorized_key:
+         user: "{{ pp_superaccount }}"
+         key: "{{ hostvars[groups['citus_coordinator'][0]].coord_key.public_key}}"
+         state: present
+         path: /var/lib/pgsql/.ssh/authorized_keys
+         comment: "postgres@{{ hostvars[groups['citus_coordinator'][0]].ansible_hostname}}"
```
Comment (tanderegg) [inline deploy/deploy_citus_server.yml]: This and `key` are missing a trailing space before the curly braces.

File: deploy/roles/citus_coordinator/templates/archive-wal-file.j2 [hunk 16]
```diff
@@ -0,0 +1,38 @@
+#!/bin/bash
+# Generated at {{ ansible_date_time.iso8601 }} for host {{ inventory_hostname }}
+# Use like this (in /var/lib/pgsql/10/data/postgresql.conf):
+# archive_command = '/var/lib/pgsql/10/data/archive-wal-file %p %f'
```
Comment (tanderegg) [inline deploy/roles/citus_coordinator/templates/archive-wal-file.j2:4]: Great script, thanks!  After this initial review, I'll try and do a run-through on Vagrant, but it looks good.

File: deploy/roles/citus_coordinator_standby/handlers/main.yml [hunk 17]
```diff
@@ -1,7 +1,14 @@
 ---
+# copied this from aurora/deploy/roles/shared/handlers/rhel7.yml
```
Comment (tanderegg) [inline deploy/roles/citus_coordinator_standby/handlers/main.yml:2]: In my other PR, I actually removed the `restart postgres` handler from the `citus` role, since it was unnecessary given that any playbook running that role is certainly using the `postgres_server` role as well, which already provides the handler.  This keeps things DRY'er.

File: deploy/roles/citus_coordinator_standby/tasks/main.yml [hunk 18]
```diff
@@ -50,7 +62,7 @@
   become: True
   file:
     dest: "{{ pp_datadir }}/{{ pg_wal }}"
-    src: "{{ pp_xlogdir }}{{ pg_wal }}"
+    src: "{{ pp_xlogdir }}/{{ pg_wal }}"
     state: link
```
Comment (tanderegg) [inline deploy/roles/citus_coordinator_standby/tasks/main.yml:68]: This should probably also have the `when: pp_datdir != pp_xlogdir` conditional.

File: deploy/roles/common/defaults/main.yml [hunk 19]
```diff
@@ -3,7 +3,7 @@
 gluster_mount_options: "defaults,acl,_netdev"
 shared_links: []
 
-skip_glusterfs: false
+skip_glusterfs: true
```
Comment (tanderegg) [inline deploy/roles/common/defaults/main.yml]: Could you explain this a bit?  It's probably fine, but I want to make sure that we're not skipping GlusterFS when we don't need to.

File: deploy/roles/gocd/tasks/main.yaml [hunk 20]
```diff
@@ -10,10 +10,10 @@
...
   tags: gocd
 
-- name: Copy local to remote
-  copy: 
-    src: "{{ gocd_agent_package_url }}"
-    dest: "{{ gocd_agent_package_path }}/{{ gocd_agent_package_name }}"
+- name: Download GoCD Agent RPM
+  shell: "curl -o {{ gocd_agent_package_path }}/{{ gocd_agent_package_name }} {{ gocd_agent_package_url }}"
+  args:
+    creates: "{{ gocd_agent_package_path }}/{{ gocd_agent_package_name }}"
```
Comment (tanderegg) [inline deploy/roles/gocd/tasks/main.yaml:16]: Was this a bug workaround?  Normally I try to use the Ansible module rather than `shell` whenever possible.

File: deploy/roles/iptables/tasks/main.yml [hunk 21]
```diff
@@ -79,5 +79,11 @@
 - name: Allow outgoing
   command: "iptables -P OUTPUT ACCEPT"
 
+# We need initscripts so we can call /sbin/service in the next task
+- name: Ensure initscripts is installed
```
Comment (tanderegg) [inline deploy/roles/iptables/tasks/main.yml:83]: This is RHEL version independent?

File: deploy/roles/log-courier/tasks/main.yml [hunk 22]
```diff
@@ -12,12 +12,12 @@
 
 - name: Install log-courier
   yum:
-    name: "{{ item }}"
+    name: [
```
Comment (tanderegg) [inline deploy/roles/log-courier/tasks/main.yml:15]: Normally I find using the `- item` syntax easier to read, was this change required for some reason?

File: deploy/roles/pem-client/tasks/main.yml [hunk 23]
```diff
@@ -7,7 +7,7 @@
   when: install_pem
 
 - name: Ensure umask is correct
-  sudo: no
+  #sudo: no
```
Comment (tanderegg) [inline deploy/roles/pem-client/tasks/main.yml]: This was here to make sure that the `deploy` users was running the umask command as itself, and not as root.

File: deploy/roles/postgresql-server/defaults/main.yml [hunk 24]
```diff
@@ -211,7 +214,7 @@ pp_wal_sender_timeout: 60s
 pp_max_replication_slots: 0
 
 # master
-pp_syncronous_standby_names: ''
+pp_synchronous_standby_names: ''
```
Comment (tanderegg) [inline deploy/roles/postgresql-server/defaults/main.yml:215]: Nice catch :)  We just need to double check that the misspelled version isn't set in `aurora-internal` somewhere.

File: deploy/roles/postgresql-server/tasks/main.yml [hunk 25]
```diff
@@ -80,6 +85,7 @@
     owner: "{{ pp_serviceaccount }}"
     group: "{{ pp_serviceaccount }}"
     mode: 0700
+  when: ansible_distribution_major_version == "hell_froze_over" # TODO: figure out correct check
```
Comment (tanderegg) [inline deploy/roles/postgresql-server/tasks/main.yml]: What's the issue here?

File: deploy/roles/postgresql-server/tasks/main.yml [hunk 26]
```diff
@@ -229,14 +241,14 @@
 
 - name: Install requirements for postgresql Ansible tasks
   yum:
-    name: "{{ item }}"
+    name: [
```
Comment (tanderegg) [inline deploy/roles/postgresql-server/tasks/main.yml:240]: Same question here as above re: list syntax

File: deploy/roles/shared/handlers/main.yml [hunk 27]
```diff
@@ -4,7 +4,7 @@
   when: (ansible_distribution == 'CentOS' or ansible_distribution == 'Red Hat Enterprise Linux') and ansible_distribution_version == 6
 
 - include_tasks: rhel7.yml
-  when: (ansible_distribution == 'CentOS' or ansible_distribution == 'Red Hat Enterprise Linux') and ansible_distribution_version == 7
+  when: (ansible_distribution == 'CentOS' or ansible_distribution == 'Red Hat Enterprise Linux') and ansible_distribution_version == "7"
```
Comment (tanderegg) [inline deploy/roles/shared/handlers/main.yml]: Normally we don't include the distribution check, although I suppose this is technically a more robust approach.  For now perhaps we should just change this to check the version in order to be consistent with other roles.

File: deploy/roles/sqitch/tasks/main.yml [hunk 28]
```diff
@@ -12,7 +12,7 @@
     - sqitch
 
 - name: Ensure umask is correct
-  sudo: no
+  #sudo: no
```
Comment (tanderegg) [inline deploy/roles/sqitch/tasks/main.yml]: Same here as above.  I think we actually created a separate `umask` role, so perhaps that should just be used?

File: deploy/roles/stata/tasks/main.yaml [hunk 29]
```diff
@@ -157,6 +157,11 @@
   tags:
     - stata
 
+- name: DEBUG - see what is in stata_install_directory
```
Comment (tanderegg) [inline deploy/roles/stata/tasks/main.yaml]: FYI, the install directory only exists in our real environment, since we don't have a local installer for Stata to use due to licesning.

File: deploy/vagrant_hosts [hunk 30]
```diff
@@ -2,32 +2,32 @@
 
 ; Shared Component
 [admin_terminal]
-admin_terminal ansible_ssh_host=10.0.1.2
+admin_terminal_0 ansible_ssh_host=10.0.1.2
```
Comment (tanderegg) [inline deploy/vagrant_hosts]: :+1: but perhaps we should start with `_1` to be consistent with the other entries (elasticsearch, file_server)

**Review by tanderegg — CHANGES_REQUESTED** [review]
@gpacix Thanks for all your work on this, overall looks great!  These are mostly comments and questions, only a few actual changes required I believe.  I will next run this locally just to make sure it all works on my machine, thanks again!

File: deploy/roles/psql-authnz/tasks/main.yml (hunk 2 above)
Comment (gpacix) [reply deploy/roles/psql-authnz/tasks/main.yml]: I've now got the entries on separate lines when there are 3 or more of them.

File: .devcontainer/Makefile (hunk 3 above)
Comment (gpacix) [reply .devcontainer/Makefile:8]: That's a good point. I'll look into it.

File: .travis.yml (hunk 6 above)
Comment (gpacix) [reply .travis.yml:12]: Yeah; I don't yet use the variable, but it shows up in the Travis build overview and greatly aided my understanding of what was going on.

File: deploy/deploy_admin_terminal.yml (hunk 8 above)
Comment (gpacix) [reply deploy/deploy_admin_terminal.yml]: This was a change I made very early; I'll try this without sudo (i.e. with `sudo: no`) and see if it works.

File: deploy/deploy_citus_server.yml (hunk 10 above)
Comment (gpacix) [reply deploy/deploy_citus_server.yml]: I agree that this isn't the clearest way to do it. I'll improve it.

File: deploy/deploy_citus_server.yml (hunk 11 above)
Comment (gpacix) [reply deploy/deploy_citus_server.yml:156]: Probably. I'll look at this as part of clarifying the `pp_install` stuff.

File: deploy/deploy_citus_server.yml (hunk 13 above)
Comment (gpacix) [reply deploy/deploy_citus_server.yml]: Agreed.

File: deploy/roles/citus_coordinator_standby/tasks/main.yml (hunk 18 above)
Comment (gpacix) [reply deploy/roles/citus_coordinator_standby/tasks/main.yml:68]: I saw that in your PR; I'll make sure this one is similar.

File: deploy/roles/common/defaults/main.yml (hunk 19 above)
Comment (gpacix) [reply deploy/roles/common/defaults/main.yml]: I'll look into this.

File: deploy/roles/gocd/tasks/main.yaml (hunk 20 above)
Comment (gpacix) [reply deploy/roles/gocd/tasks/main.yaml:16]: Yes; this is a workaround for the "Failed to validate the SSL certificate" errors we were getting; it's simpler than trying to use Python 2.7.  It (separating download from install) pops up a couple other places, too, for the same reason.

File: deploy/roles/iptables/tasks/main.yml (hunk 21 above)
Comment (gpacix) [reply deploy/roles/iptables/tasks/main.yml:83]: I'm 99% certain it is, because I've run the `iptables` role dozens of times on both CentOS 6.x and CentOS 7.x.  Do we need more assurance than that?

File: deploy/roles/log-courier/tasks/main.yml (hunk 22 above)
Comment (gpacix) [reply deploy/roles/log-courier/tasks/main.yml:15]: Yes: Ansible gives a deprecation warning, and says the `item` syntax will not be supported in Ansible 2.10.  I took Daniel's advice to make it more readable by putting each package in the list on its own line.

File: deploy/roles/pem-client/tasks/main.yml (hunk 23 above)
Comment (gpacix) [reply deploy/roles/pem-client/tasks/main.yml]: I'll revert it and test that, then.

File: deploy/roles/postgresql-server/tasks/main.yml (hunk 26 above)
Comment (gpacix) [reply deploy/roles/postgresql-server/tasks/main.yml:240]: (See my previous reply about the deprecation warning.)

File: deploy/roles/shared/handlers/main.yml (hunk 27 above)
Comment (gpacix) [reply deploy/roles/shared/handlers/main.yml]: Probably. Maybe common or shared should do a check that it's RHEL or CentOS, and give a warning (or even error) if it's not.

File: deploy/vagrant_hosts (hunk 30 above)
Comment (gpacix) [reply deploy/vagrant_hosts]: This should be an easy fix: it's very rare that anything refers to a specific host; almost all references are by group.

File: deploy/deploy_admin_terminal.yml (hunk 8 above)
Comment (gpacix) [reply deploy/deploy_admin_terminal.yml]: I've reinstated these and updated them to `become: no`.

File: deploy/deploy_citus_server.yml (hunk 13 above)
Comment (gpacix) [reply deploy/deploy_citus_server.yml]: In every role or playbook this branch touches, I've made sure there's a space after `{{` and a space before `}}`.

File: deploy/deploy_citus_server.yml (hunk 15 above)
Comment (gpacix) [reply deploy/deploy_citus_server.yml]: (see previous)

File: deploy/roles/citus_coordinator_standby/tasks/main.yml (hunk 18 above)
Comment (gpacix) [reply deploy/roles/citus_coordinator_standby/tasks/main.yml:68]: Added the `when` conditional.

File: deploy/roles/common/defaults/main.yml (hunk 19 above)
Comment (gpacix) [reply deploy/roles/common/defaults/main.yml]: I accidentally committed that, and have now changed it back to `false`.  (It didn't affect any of the Travis build's jobs, since they all set `skip_glusterfs` to `true`, except for Job 3, which doesn't use GlusterFS at all anyway.)

File: deploy/roles/pem-client/tasks/main.yml (hunk 23 above)
Comment (gpacix) [reply deploy/roles/pem-client/tasks/main.yml]: Reinstated (see above reply for more detail).

File: deploy/roles/shared/handlers/main.yml (hunk 27 above)
Comment (gpacix) [reply deploy/roles/shared/handlers/main.yml]: Dropped the `ansible_distribution` clauses, changed `ansible_distribution_version` to `ansible_distribution_major_version`, and fixed the unquoted 6 a couple lines above.

File: deploy/roles/sqitch/tasks/main.yml (hunk 28 above)
Comment (gpacix) [reply deploy/roles/sqitch/tasks/main.yml]: I'm not clear on why we're changing the deploy user's `.bashrc` to set the `umask`.  I reinstated the `sudo: no` as `become: no` (see above).  It might be good to use the `umask` role, but that's a bigger change and I figured I'd go with the quicker fix right now.

File: deploy/roles/stata/tasks/main.yaml (hunk 29 above)
Comment (gpacix) [reply deploy/roles/stata/tasks/main.yaml]: Thanks.  I've removed the DEBUG task.

File: deploy/vagrant_hosts (hunk 30 above)
Comment (gpacix) [reply deploy/vagrant_hosts]: Narrator: it actually **was** an easy fix.

File: archive-wal-file (hunk 7 above)
Comment (gpacix) [reply archive-wal-file]: Yes; deleted this file at some point.

File: deploy/roles/postgresql-server/tasks/main.yml (hunk 25 above)
Comment (gpacix) [reply deploy/roles/postgresql-server/tasks/main.yml]: That was for developing; I've removed the "when" clause now.

**Review by tanderegg — APPROVED** [review]
